dohq_artifactory/admin.py

```diff
@@ -686,34 +690,128 @@ def _read_response(self, response):
         self.includesPattern = response["includesPattern"]
         self.excludesPattern = response["excludesPattern"]
         self._repositories = response.get("repositories", [])
-        if "principals" in response:
-            if "users" in response["principals"]:
-                self._users = response["principals"]["users"]
-            if "groups" in response["principals"]:
-                self._groups = response["principals"]["groups"]
+        self._users = {}
+        self._groups = {}
+        if response.get("principals"):
+            if response["principals"].get("users"):
+                self._users = self._permissions_from_json(
+                    response["principals"]["users"]
+                )
+            if response["principals"].get("groups"):
+                self._groups = self._permissions_from_json(
+                    response["principals"]["groups"]
+                )
 
-    def add_repository(self, *args):
-        self._repositories.extend([x if isinstance(x, str) else x.name for x in args])
+    @classmethod
+    def _principal_parse(cls, name, permissions):
+        return cls._principal_name_parse(name), cls._permissions_parse(permissions)
 
-    @staticmethod
-    def _add_principals(name, permissions, principals):
-        if isinstance(permissions, str):
-            permissions = [permissions]
-        permissions = list(set(permissions))
+    @classmethod
+    def _permissions_from_json(cls, permissions_map):
+        result = {}
+        for key, permissions in permissions_map.items():
+            name, new_permissions = cls._principal_parse(key, permissions)
+            result[name] = new_permissions
+        return result
+
+    @classmethod
+    def _principal_name_parse(cls, name):
         if isinstance(name, AdminObject):
             name = name.name
-        principals[name] = permissions
+        return name
+
+    @classmethod
+    def _permissions_parse(cls, permissions):
+        if isinstance(permissions, str):
+            permissions = re.sub(r"\W", "", permissions.strip()).split("")
+        permissions = list(set(permissions))
+
+        for permission in permissions:
+            if permission not in cls.ROLE_ADMIN:
+                raise ValueError("Unknown permission {name}".format(name=permission))
+        return permissions
 
     def add_user(self, name, permissions):
-        self._add_principals(name, permissions, self._users)
+        name, permissions = self._principal_parse(name, permissions)
+        self._users[name] = permissions
+
+    def remove_user(self, *args):
```

Comment:
We can name arguments as we want, really. We tends to name `args kwargs` in python, but the below example is more readable. What do you think?
```python
def remove_users(self, *users):
    ... 
```